figures/filters.py

```diff
@@ -391,6 +395,69 @@ def filter_enrolled_in_course_id(self, queryset,
         return queryset.filter(id__in=user_ids).using(read_replica_or_default())
 
 
+class CustomLearnerSearchFilter(filters.SearchFilter):
+    def filter_queryset(self, request, queryset, view):
+
+        search_terms = self.get_search_terms(request)
+
+        if not search_terms:
+            return queryset
+
+        search_queries = Q()
+        for term in search_terms:
+            search_queries |= (
+                Q(original_username__icontains=term) | 
+                Q(original_email__icontains=term)
+            )
+
+        retired_users = queryset.filter(email__startswith=settings.RETIRED_EMAIL_PREFIX)
+
+        retirement_statuses = UserRetirementStatus.objects.filter(
+            user__in=retired_users
+        ).filter(search_queries)
+
+        return super().filter_queryset(request, queryset, view) | queryset.filter(
+            id__in=retirement_statuses.values_list('user_id', flat=True)
+            )
+
+class NullsLastOrderingFilter(filters.OrderingFilter):
+
+    def filter_queryset(self, request, queryset, view):
+        ordering = self.get_ordering(request, queryset, view)
+        if not ordering:
+            return queryset
+        
+        annotations = {}
+        ordering_expressions = []
+        
+        for field in ordering:
+            clean_field = field.lstrip("-")
+            field_parts = clean_field.split('__')
+            
+            if len(field_parts) > 1:
+                annotation_name = f"{field_parts[0]}_{field_parts[1]}_non_empty"
```

Comment:
Will it work? 
`annotation_name = f"{clean_field.replace('__', '_')}_non_empty"`